filebeat/input/journald/pkg/journalctl/reader.go

```diff
@@ -0,0 +1,299 @@
+	}
+
+	return args
+}
+
+// New instantiates and starts a reader for journald logs.
+//
+// The Reader starts a `journalctl` process with JSON ouput to read the journal
+// entries. Units and syslog identifiers are passed using the corresponding CLI
+// flags, matchers are passed directly to `journalctl` then transports are added
+// as matchers using `_TRANSPORTS` key.
+//
+// `mode` defines the 'seek mode'. It indicates whether the journal should be
+// read from the tail, head or starting from the cursor. If a cursor is passed,
+// then the seek mode is ignored.
+//
+// To start reading from a relative time, use mode: SeekSince and since should
+// be a time.Duration relative to the current time to start reading the
+// journald.
+//
+// File is the journal file to be read, for the system journal use the string
+// `LOCAL_SYSTEM_JOURNAL`.
+//
+// It's the caller's responsibility to call `Close` on the reader to stop
+// the `journalctl` process.
+//
+// If `canceler` is cancelled, the reading goroutine is stopped and subsequent
+// calls to `Next` will return an error.
+func New(
+	logger *logp.Logger,
+	canceler input.Canceler,
+	units []string,
+	syslogIdentifiers []string,
+	transports []string,
+	matchers journalfield.IncludeMatches,
+	mode SeekMode,
+	cursor string,
+	since time.Duration,
+	file string) (*Reader, error) {
+
+	args := []string{"--utc", "--output=json", "--follow"}
+	if file != "" && file != localSystemJournalID {
+		args = append(args, "--file", file)
+	}
+
+	args = handleSeekAndCursor(args, mode, since, cursor)
+
+	for _, u := range units {
+		args = append(args, "--unit", u)
+	}
+
+	for _, i := range syslogIdentifiers {
+		args = append(args, "--identifier", i)
+	}
+
+	for _, m := range matchers.Matches {
+		args = append(args, m.String())
+	}
+
+	for _, m := range transports {
+		args = append(args, fmt.Sprintf("_TRANSPORT=%s", m))
+	}
+
+	logger.Debugf("Journalctl command: journalctl %s", strings.Join(args, " "))
+	cmd := exec.Command("journalctl", args...)
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return &Reader{}, fmt.Errorf("cannot get stdout pipe: %w", err)
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return &Reader{}, fmt.Errorf("cannot get stderr pipe: %w", err)
+	}
+
+	r := Reader{
+		cmd:      cmd,
+		dataChan: make(chan []byte),
+		errChan:  make(chan error),
+		logger:   logger,
+		stdout:   stdout,
+		stderr:   stderr,
+		canceler: canceler,
+	}
+
+	// Goroutine to read errors from stderr
+	r.wg.Add(1)
+	go func() {
+		defer r.logger.Debug("stderr goroutine done")
+		defer r.wg.Done()
+		reader := bufio.NewReader(r.stderr)
+		msgs := []string{}
+		for {
+			line, err := reader.ReadString('\n')
+			if errors.Is(err, io.EOF) {
+				if len(msgs) == 0 {
+					return
+				}
+				errMsg := fmt.Sprintf("Journalctl wrote errors: %s", strings.Join(msgs, "\n"))
+				logger.Errorf(errMsg)
+				r.errChan <- errors.New(errMsg)
+				return
+			}
+			msgs = append(msgs, line)
```

Comment:
There is no size limit here. It seems safer to print the lines as you receive them until waiting for io.EOF.

Reply:
That's a very good point, which made me realise a even bigger problem: the current implementation never reads from the `r.errChan` :man_facepalming:.

Regardless, once we start reading from it when do we stop? What is the break condition once we have data coming from `stderr`. I'll read again the `jouranlctl` documentation and see if I can find something helpful to have a breaking condition there.